### ISSUE TITLE
Set grad to None if grad.sum() == 0

### ISSUE THREAD

**Issue by CanyonWind** [megvii-model/DetNAS#3]
Hi, great work!

I am quite new in PyTorch, so please forgive me if this is just some common approach in PyTorch. Could you please educate me on why the `grad` is set to `None` when `grad.sum() == 0 `.

https://github.com/megvii-model/DetNAS/blob/94623fd3c65934d1fba976c5c5d9d7f6b855ea90/Supernet-ImageNet/train.py#L229-L231

The branches which not have been used in each batch will have 0 gradients, right? Then what's the purpose of setting them to `None`, and how it will benefit the training?

Thanks

**Comment by yukang2017** [comment]
Hi, this is a good question. We do so to avoid the unwanted parameter update.

When training the supernet, we only want to update the parameters on the path we sampled in each iteration. However, parameters on other paths would be updated by `momentum` and `weight_decay`, if `not p.grad is None`.

You can also refer to the inner code of `optimizer.step()` in the torch.

**Comment by CanyonWind** [comment]
It makes good sense. Thanks!
